qiskit_experiments/library/randomized_benchmarking/rb_experiment.py

```diff
@@ -133,86 +141,128 @@ def circuits(self) -> List[QuantumCircuit]:
-            circ_op = circ_op.compose(group_elt_op)
-            for circ in circs:
-                circ.append(group_elt_gate, qubits)
+            for elem in seq:
+                circ.append(self._to_instruction(elem), qubits)
                 circ.barrier(qubits)
-            if current_length + 1 in lengths:
-                # copy circuit and add inverse
-                inv = circ_op.adjoint()
-                rb_circ = circs.pop()
-                rb_circ.append(inv, qubits)
-                rb_circ.barrier(qubits)
-                rb_circ.metadata = {
-                    "experiment_type": self._type,
-                    "xval": current_length + 1,
-                    "group": "Clifford",
-                    "physical_qubits": self.physical_qubits,
-                }
-                rb_circ.measure_all()
-                circuits.append(rb_circ)
+
+            # Compute inverse, compute only the difference from the previous shorter sequence
+            for elem in seq[len(prev_seq) :]:
+                prev_elem = self.__compose_clifford(prev_elem, elem)
+            prev_seq = seq
+            inv = self.__adjoint_clifford(prev_elem)
+
+            circ.append(self._to_instruction(inv), qubits)
+            circ.barrier(qubits)  # TODO: Can we remove this? (measure_all inserts one more barrier)
+            circ.measure_all()
+            circuits.append(circ)
         return circuits
 
+    def _sample_sequence(self, length: int, rng: Generator) -> Sequence[SequenceElementType]:
+        """Sample a RB sequence with the given length.
+
+        Args:
+            length: RB sequences length.
+            rng: Generator object for random number generation.
+
+        Returns:
+            A RB sequence.
+        """
+        # Return integer instead of Clifford object for 1 or 2 qubit case
+        if self.num_qubits == 1:
+            return rng.integers(24, size=length)
+        if self.num_qubits == 2:
+            return rng.integers(11520, size=length)
+
+        return [random_clifford(self.num_qubits, rng) for _ in range(length)]
+
+    def _to_instruction(self, elem: SequenceElementType) -> Instruction:
+        # TODO: basis transformation in 1Q (and 2Q) cases for speed
+        # Switching for speed up
+        if isinstance(elem, Integral):
+            if self.num_qubits == 1:
+                return _clifford_1q_int_to_instruction(elem)
+            if self.num_qubits == 2:
+                return _clifford_2q_int_to_instruction(elem)
+        return elem.to_instruction()
+
+    def __identity_clifford(self) -> SequenceElementType:
+        if self.num_qubits <= 2:
+            return 0
+        return Clifford(np.eye(2 * self.num_qubits))
+
+    def __compose_clifford(
+        self, lop: SequenceElementType, rop: SequenceElementType
+    ) -> SequenceElementType:
+        # TODO: Speed up 1Q (and 2Q) cases using lookup table
+        if self.num_qubits == 1:
+            if isinstance(lop, Integral):
```

Comment:
Can we move this type check to `_sequences_to_circuits`? I don't think there can be a mixture of Clifford and Integral  type in the sequence. I think this can be called only when the sequence is Integral type, namely <3Q.

Reply:
I see why it is implemented like this. Interleaved RB will inject Clifford object into Integral sequence. Seems like this is okey for now, but I guess this could become dominant after circuit generation logic is optimized. Perhaps we can convert interleaved element into Integral and prohibit from generating a sequence of mixed type. Please write some comment if you keep this as-is.

Reply:
Right, we're in the middle of performance optimization and we need to handle mixture of integer Clifford and standard Clifford for now. Those implementation would be updated in the future performance improvement PRs. I'll add some comment.